### ISSUE TITLE
Async over sync in SqlTransaction.CommitAsync

### ISSUE THREAD

**Issue by JulianRooze** [dotnet/SqlClient#2215]
### Describe the bug

We are currently load testing our application and are encountering thread pool starvation with a lot of threads waiting on this stack trace:

```
Thread (0xF6DA):
  [Native Frames]
  System.Private.CoreLib!System.Threading.ManualResetEventSlim.Wait(int32,value class System.Threading.CancellationToken)
  System.Net.Sockets!System.Net.Sockets.SocketAsyncContext.PerformSyncOperation(value class OperationQueue`1<!!0>&,!!0,int32,int32)
  System.Net.Sockets!System.Net.Sockets.SocketAsyncContext.ReceiveFrom(value class System.Memory`1<unsigned int8>,value class System.Net.Sockets.SocketFlags&,unsigned int8[],int32&,int32,int32&)
  System.Net.Sockets!System.Net.Sockets.SocketPal.Receive(class System.Net.Sockets.SafeSocketHandle,unsigned int8[],int32,int32,value class System.Net.Sockets.SocketFlags,int32&)
  System.Net.Sockets!System.Net.Sockets.Socket.Receive(unsigned int8[],int32,int32,value class System.Net.Sockets.SocketFlags,value class System.Net.Sockets.SocketError&)
  System.Net.Sockets!System.Net.Sockets.NetworkStream.Read(unsigned int8[],int32,int32)
  System.Private.CoreLib!System.IO.Stream.Read(value class System.Span`1<unsigned int8>)
  System.Net.Sockets!System.Net.Sockets.NetworkStream.Read(value class System.Span`1<unsigned int8>)
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.SNI.SslOverTdsStream.Read(value class System.Span`1<unsigned int8>)
  System.Net.Security!System.Net.Security.SslStream+<EnsureFullTlsFrameAsync>d__157`1[System.Net.Security.SyncReadWriteAdapter].MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start(!!0&)
  System.Net.Security!System.Net.Security.SslStream.EnsureFullTlsFrameAsync(value class System.Threading.CancellationToken)
  System.Net.Security!System.Net.Security.SslStream+<ReadAsyncInternal>d__159`1[System.Net.Security.SyncReadWriteAdapter].MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start(!!0&)
  System.Net.Security!System.Net.Security.SslStream.Read(unsigned int8[],int32,int32)
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.SNI.SNITCPHandle.Receive(class Microsoft.Data.SqlClient.SNI.SNIPacket&,int32)
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.SNI.TdsParserStateObjectManaged.ReadSyncOverAsync(int32,unsigned int32&)
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.TdsParserStateObject.ReadSniSyncOverAsync()
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.TdsParserStateObject.TryReadNetworkPacket()
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.TdsParserStateObject.TryPrepareBuffer()
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.TdsParserStateObject.TryReadByte(unsigned int8&)
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.TdsParser.TryRun(value class Microsoft.Data.SqlClient.RunBehavior,class Microsoft.Data.SqlClient.SqlCommand,class Microsoft.Data.SqlClient.SqlDataReader,class Microsoft.Data.SqlClient.BulkCopySimpleResultSet,class Microsoft.Data.SqlClient.TdsParserStateObject,bool&)
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.TdsParser.Run(value class Microsoft.Data.SqlClient.RunBehavior,class Microsoft.Data.SqlClient.SqlCommand,class Microsoft.Data.SqlClient.SqlDataReader,class Microsoft.Data.SqlClient.BulkCopySimpleResultSet,class Microsoft.Data.SqlClient.TdsParserStateObject)
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.TdsParser.TdsExecuteTransactionManagerRequest(unsigned int8[],value class TransactionManagerRequestType,class System.String,value class TransactionManagerIsolationLevel,int32,class Microsoft.Data.SqlClient.SqlInternalTransaction,class Microsoft.Data.SqlClient.TdsParserStateObject,bool)
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.SqlInternalConnectionTds.ExecuteTransaction2005(value class TransactionRequest,class System.String,value class System.Data.IsolationLevel,class Microsoft.Data.SqlClient.SqlInternalTransaction,bool)
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.SqlInternalTransaction.Commit()
  Microsoft.Data.SqlClient!Microsoft.Data.SqlClient.SqlTransaction.Commit()
  System.Data.Common!System.Data.Common.DbTransaction.CommitAsync(value class System.Threading.CancellationToken)
```

The way I read is that `SqlTransaction.CommitAsync` is not overridden but calls into `DbTransaction.CommitAsync` which itself simply delegates to the sync implementation of `Commit`. 

I understand that committing a transaction shouldn't normally take a long time (and is likely caused by database load), but in the event that it does it exacerbates application responsiveness issues due to thread starvation.

### To reproduce

Any call to a `SqlTransaction.CommitAsync()`.

### Expected behavior

All I/O is done asynchronously. 

### Further technical details
Microsoft.Data.SqlClient version: 5.1.2
.NET target: .NET 7.0.403

**Comment by roji** [comment]
Just to be precise, the above isn't sync-over-async, but rather async-over-sync.

However, the use of synchronous I/O still blocks the thread, and if it happens on the thread pool, can indeed cause starvation issues. In general, it should be possible to use SqlClient in a fully asynchronous manner, without ever blocking the thread.

**Comment by JulianRooze** [comment]
@roji I updated the title 😄 I get the two mixed up, and doubted it here as well but based my final decision on the method you see in the above stack trace where it calls `TdsParserStateObjectManaged.ReadSyncOverAsync`.

**Comment by roji** [comment]
Yeah, you may be right - I don't know anything about the internals, but the method does suggest that sync-over-async occurs internally - which would also not be good.

**Comment by maffelbaffel** [comment]
I am currently also working on a thread starvation issue in one of our services.
This service is kinda write heavy on sql inserts.

Whenever there is more network latency than usual, we see increasing thread starvation.
We should be able to resolve this issue by introducing a caching layer and reduce writes. 

Still this is very unfortunate for a official sql client to have blocking calls in `CommitAsync()`

**Comment by David-Engel** [comment]
Because CommitAsync() is not implemented in SqlClient. There is no code for there to be a bug in. System.Data.Common.DbTransaction's CommitAsync() is calling Commit() when a provider doesn't have its own CommitAsync() implementation.

**Comment by jtheisen** [comment]
@David-Engel I would argue that while a completely missing asynchronous implementation isn't a bug, an incomplete implementation is, unless it's documented and people don't run into it unexpectedly. It's unexpected now and causes issues such as frozen user interfaces on rollbacks.

Arguably the `Db`* base classes should have a boolean property indicating whether there actually is such a (complete) asynchronous implementation, or the entire class hierarchy should have been extended differently (with separate interfaces for the asynchronous stuff).

I don't want to be too annoying here, but it's more than just a missing feature as currently it's suggesting it has that feature although it doesn't implement it fully.

I would also assume that this isn't intentional but merely an oversight during the implementation.

In my own case, the workaround is simply to run the queries on a different thread, but that requires the user knows that this issue exists in the first place.

**Comment by roji** [comment]
@jtheisen all async APIs in ADO.NET delegate to the synchronous implementations by default; this is also true of e.g. DbConnection.OpenAsync() and DbCommand.ExecuteReader(); this pattern was adopted when async APIs were first retrofitted into ADO.NET (which originally only had sync APIs, having been designed before async even existed).

I'm not sure whether I'd have adopted this pattern today, but that's not really relevant at this point - that's the standard behavior across all ADO.NET APIs. I don't think this can be considered a bug - an async method **can** block for various reasons, even though that's certainly not the intended way for such methods to work.

In any case, regardless of how you look at this, the way forward here is for SqlClient to implement CommitAsync() (and RollbackAsync()) in the proper, async way, and that's already tracked by #113... I don't see anything beyond that in this issue, which can probably be closed.

**Comment by David-Engel** [comment]
Closing as this will be addressed when #113 is implemented.
